File: 692. Top K Frequent Words/solution.py
from collections import defaultdict

class Solution:
    def topKFrequent(self, words, k):
        """
        :type words: List[str]
        :type k: int
        :rtype: List[str]
        """
        
        # count the frequency of each word
        freq = defaultdict(lambda: 0)
        for word in words:
            freq[word] += 1
            
        # sort the tuple using the freq first and then the word
        sortFreq = [(w, f) for w, f in freq.items()]
        sortFreq.sort(key=lambda x: (-x[1], x[0]))
        
        # grab the first k words 
        return [w[0] for w in sortFreq[:k]]

File: 692. Top K Frequent Words/solution.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        // get the frequency
        Map<String, Integer> freq = new HashMap<>();
        for (String word: words) {
            freq.put(word, freq.getOrDefault(word, 0)+1);
        }
        
        // sort them
        List<Node> nodes = new ArrayList<>();
        freq.forEach((u, v) -> nodes.add(new Node(v, u)));
        Collections.sort(nodes, new Comparator<Node>(){
            public int compare(Node n1, Node n2) {
                if (n1.f == n2.f) {
                    return n1.s.compareTo(n2.s);
                } else {
                    return n2.f - n1.f;
                }
            }
        });
        
        // final solution
        List<String> sol = new ArrayList<String>();
        for (int i = 0; i < k; i++) {
            sol.add(nodes.get(i).s);
        }
        
        return sol;
    }
    
    public class Node {
        public int f;
        public String s;
        public Node(int f, String s) {
            this.f = f;
            this.s = s;
        }
    }
}
